src/sqlfluff/rules/L042.py

```diff
@@ -218,8 +217,67 @@ def _calculate_fixes(
     return lint_results
 
 
+def _get_first_select_statement_descendant(
+    segment: BaseSegment,
+) -> Optional[BaseSegment]:
+    """Find first SELECT statement segment (if any) in descendants of 'segment'."""
+    for select_statement in segment.recursive_crawl(
+        "select_statement", recurse_into=False
+    ):
+        # We only want the first one.
+        return select_statement
+    return None
+
+
+def _get_sources_from_select(segment: BaseSegment, dialect: Dialect) -> Set[str]:
+    """Given segment, return set of table or alias names it queries from."""
+    result = set()
+    select = None
+    if segment.is_type("select_statement"):
+        select = segment
+    elif segment.is_type("with_compound_statement"):
+        # For WITH statement, process the main query underneath.
+        select = _get_first_select_statement_descendant(segment)
+    if select and select.is_type("select_statement"):
+        select_info = get_select_statement_info(select, dialect)
+        if select_info:
+            for a in select_info.table_aliases:
+                # For each table in FROM, return table name and any alias.
+                if a.ref_str:
+                    result.add(a.ref_str)
+                if a.object_reference:
+                    result.add(a.object_reference.raw)
+    return result
+
+
+def _is_correlated_subquery(
+    nested_select: Segments, select_source_names: Set[str], dialect: Dialect
+):
+    """Given nested select and the sources of its parent, determine if correlated.
+
+    https://en.wikipedia.org/wiki/Correlated_subquery
+    """
+    if not nested_select:
+        return False
+    select_statement = _get_first_select_statement_descendant(nested_select[0])
+    if not select_statement:
+        return False
+    nested_select_info = get_select_statement_info(select_statement, dialect)
+    if nested_select_info:
+        for r in nested_select_info.reference_buffer:
+            for tr in r.extract_possible_references(  # type: ignore
+                level=r.ObjectReferenceLevel.TABLE  # type: ignore
```

Comment:
Used `type: ignore` because Python doesn't know `r` is a `ObjectReferenceSegment` or related type.